src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndex.java

```diff
@@ -1,34 +1,66 @@
+/*
+ *   Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
 package com.amazon.opendistroforelasticsearch.knn.index;
 
+import java.io.File;
+import java.io.IOException;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
 
-import com.amazon.opendistroforelasticsearch.knn.index.util.KNNEngine;
+/**
+ * Encapsulates functionality to interface with an underlying k-NN Engine
+ */
+public abstract class KNNIndex implements AutoCloseable {
 
-public abstract class KNNIndex implements AutoCloseable{
+    protected volatile boolean isClosed = false;
+    protected final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
 
+    protected final long indexPointer;
+    protected final long indexSize;
+    protected final SpaceType spaceType;
 
-    public KNNIndex() {}
```

Comment:
you don't need?
```
protected KNNIndex() {}
```

Reply:
No I dont think so. The constructor for the index is only needed in the KNNFaissIndex and KNNNmslibIndex constructors, which are only used in their respective loadIndex static functions. 

What do you think?

Reply:
Got it. if it is not used anywhere, then we don't have to add it.